File: daily/2024/2563_Count_the_Number_of_Fair_Pairs.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        long long ans = 0;
        int start = n - 1;
        int end = n - 1;
        for (int i = 0; i < n; i++){
            if (nums[i] + nums[start] < lower) continue;
            while(start - 1 >= 0 && nums[start - 1] + nums[i] >= lower) start--;
            while(end - 1 >= 0 && nums[end] + nums[i] > upper) end--;
            if (i >= end) return ans;
            if (i < start) ans += end - start + 1;
            else ans += end - i;
            // cout << nums[i] << " : " << nums[start] << " - " << nums[end] << endl;
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {0,1,7,4,4,5};
    int lower = 3;
    int upper = 6;
    // vector<int> nums = {1,7,9,2,5};
    // int lower = 11;
    // int upper = 11;
    // vector<int> nums = {0,0,0,0,0,0};
    // int lower = 0;
    // int upper = 0;
    Solution S;
    cout << S.countFairPairs(nums, lower, upper) << endl;
    return 0;
}

File: daily_python/2563_Count_the_Number_of_Fair_Pairs.py
import os
import sys
from typing import List

class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        n = len(nums)
        start = n - 1
        end = n - 1
        ans = 0
        for i in range(n - 1):
            if nums[i] + nums[start] < lower:
                continue
            while start - 1 >= 0 and nums[i] + nums[start - 1] >= lower:
                start -= 1
            while end - 1 >= 0 and nums[i] + nums[end] > upper:
                end -= 1
            if i >= end:
                return ans
            start = max(start, i + 1)
            ans += end - start + 1
        return ans
    
def readDatSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            lower = int(lines[1].split('=')[1].strip()[:-1])
            upper = int(lines[2].split('=')[1].strip()[:-1])
            dataset.append((nums, lower, upper))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDatSet(filename)
    results = []
    solution = Solution()
    for nums, lower, upper in dataset:
        results.append(solution.countFairPairs(nums, lower, upper))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')